# Data Gathering

In [1]:
import pandas as pd
import json
from timeit import default_timer as timer
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

### 1. Create a DF out of the desired data from the JSON file created using the Twitter API

In [2]:
tweet_data = []
#Open and read each line of the Json file. 
with open("tweet_json.txt", 'r') as json_file:
    for line in json_file:
        json_data = json.loads(line)
        timestamp = json_data['created_at'] #Pull out 4 data points (timestamp, tweet_id, likes, and retweets)
        likes = json_data['favorite_count']
        tweet_id = json_data['id']
        retweets = json_data['retweet_count']
        
        tweet_data.append({'timestamp': timestamp, #Append those 4 pieces of data to a list in the form of a dict (key/value pairs) to be converted later into a DF
                         'likes': likes,
                         'tweet_id': tweet_id,
                         'retweets': retweets})


In [3]:
tweet_data_df = pd.DataFrame(tweet_data, columns = ['timestamp', 'likes', 'tweet_id', 'retweets'])#Create the dataframe with the JSON data above and specific column names

In [4]:
tweet_data_df.head()

timestamp  likes            tweet_id  retweets
0  Tue Aug 01 16:23:56 +0000 2017  35562  892420643555336193      7534
1  Tue Aug 01 00:17:27 +0000 2017  30753  892177421306343426      5589
2  Mon Jul 31 00:18:03 +0000 2017  23130  891815181378084864      3703
3  Sun Jul 30 15:58:51 +0000 2017  38869  891689557279858688      7717
4  Sat Jul 29 16:00:24 +0000 2017  37129  891327558926688256      8316

In [5]:
tweet_data_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   timestamp  2331 non-null   object
 1   likes      2331 non-null   int64 
 2   tweet_id   2331 non-null   int64 
 3   retweets   2331 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 73.0+ KB


### 2. Create a DF from the provided WeRateDogs twitter archive data csv

In [6]:
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')

In [7]:
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [8]:
twitter_archive.in_reply_to_status_id.isnull().value_counts()

True     2278
False      78
Name: in_reply_to_status_id, dtype: int64

### 3. Create a DF utilizing the requests package to work with the provided image predictions url

In [9]:
import requests
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

In [10]:
r = requests.get(url)

In [11]:
with open('image_predictions.tsv', 'wb') as file:
    file.write(r.content)

In [12]:
image_predictions = pd.read_csv('image_predictions.tsv', sep='\t')
image_predictions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [13]:
image_predictions.sample(3)

tweet_id                                            jpg_url  \
190   669393256313184256    https://pbs.twimg.com/media/CUopnHPVEAAcL2o.jpg   
1705  817423860136083457  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1136  728760639972315136    https://pbs.twimg.com/media/Ch0T71OWMAA4yIw.jpg   

      img_num              p1   p1_conf  p1_dog                p2   p2_conf  \
190         1  cocker_spaniel  0.359843    True  Blenheim_spaniel  0.139519   
1705        1        ice_bear  0.336200   False           Samoyed  0.201358   
1136        1        Pembroke  0.939134    True          Cardigan  0.054336   

      p2_dog          p3   p3_conf  p3_dog  
190     True  toy_poodle  0.132746    True  
1705    True  Eskimo_dog  0.186789    True  
1136    True   Chihuahua  0.005590    True

#### Now that the DFs have been created, a copy of each is generated for future use during assessment or cleaning stages of the Data Wrangling process

In [14]:
tweet_data_df_clean = tweet_data_df.copy()
image_predictions_clean = image_predictions.copy()
twitter_archive_clean = twitter_archive.copy()

# Assess

### 1. Programmatic and Visual Assessment

#### Quality
<ul> 
    <li>tweet_data_df_clean:
        <ol>
            <li>timestamp data provides more data than necessary</li>
            <li>there are 163 tweets with 0 likes, but all of them have at least 19 or more retweets</li>
            <li>timestamp is datatype is an object and it should be datetime</li>
            <li>tweet_id datatype is int64 and it should be a string</li>
        </ol>
    </li>
    <li>image_predictions_clean
        <ol>
            <li>inconsistent formatting for the names potential image matches</li>
            <li>there are many image predictions (even in p1) which are not dogs, sometimes not even animals</li>
            <li>there are 64 image predictions where the first two predictions are not dogs</li>
            <li>there are 324 image predictions where all three predictions are not dogs</li>
            tweet_id datatype is int64 and it should be a string
        </ol>
    </li>
    <li>twitter_archive_clean
        <ol>
            <li>there are more data entries in twitter_archive_clean than there are in both image_predictions_clean and tweet_data_df_clean, meaning there are tweets which do not have images associated with them in the image database, nor retweet/like information</li>
            <li>there are null values for in_reply_to_status_id, in_reply_to_user_id, retweeted_status_id, retweeted_status_user_id, and retweeted_status_timestamp</li>
            <li>there are 181 retweets as identified by retweeted_status_id, retweeted_status_user_id, and retweeted_status_timestamp columns</li>
            <li>there are 78 replies as identfied by in_reply_to_status_id and in_reply_to_user_id columns</li>
            <li>tweet_id datatype is int64 and it should be a string</li>
            <li>there are 59 rows missing expanded_urls data</li>
            <li>timestamp is datatype is an object and it should be datetime</li>
            <li>inaccurate rating_numerator and rating_denominator informatin</li>
        </ol>
    </li>
</ul>        

#### Tidiness
<ul> 
    <li>tweet_data_df_clean:
        <ol>
            <li>tweet_data_df_clean does not need to be in a separate table and should be merged with twitter_archive_clean, the larger complete df</li>
        </ol>
    </li>
    <li>image_predictions_clean
        <ol>
            <li>image_predictions_clean data could be condensed and combined with twitter_archive_clean, though it also may be better off on its own due to the nature of its content, image identification vs basic tweet information. </li>
        </ol>
    </li>
    <li>twitter_archive_clean
        <ol>
            <li>dog stages are in four different columns</li>
            <li>source and expanded url provide the same information</li>
        </ol>
    </li>
</ul>        

## tweet_data_df_clean

In [15]:
tweet_data_df_clean.sort_values(by = 'retweets').head(10)

timestamp  likes            tweet_id  retweets
279   Sat Mar 04 17:56:49 +0000 2017    138  838085839343206401         1
1271  Thu Mar 10 17:35:20 +0000 2016     49  707983188426153984         2
262   Sat Mar 11 22:59:09 +0000 2017    171  840698636975636481         2
328   Thu Feb 16 04:45:50 +0000 2017     62  832088576586297345         2
109   Fri Jun 02 19:38:25 +0000 2017    113  870726314365509632         3
29    Sat Jul 15 16:51:35 +0000 2017    110  886267009285017600         4
1056  Sat Jun 04 00:32:32 +0000 2016    106  738891149612572673         6
54    Sun Jul 02 21:58:53 +0000 2017    115  881633300179243008         7
411   Tue Jan 17 00:33:26 +0000 2017    254  821153421864615936        10
63    Tue Jun 27 12:14:36 +0000 2017    291  879674319642796034        10

In [16]:
tweet_data_df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   timestamp  2331 non-null   object
 1   likes      2331 non-null   int64 
 2   tweet_id   2331 non-null   int64 
 3   retweets   2331 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 73.0+ KB


In [17]:
tweet_data_df_clean.describe()

likes      tweet_id      retweets
count    2331.00000  2.331000e+03   2331.000000
mean     7430.58387  7.419079e+17   2648.574432
std     11539.21545  6.823170e+16   4479.174285
min         0.00000  6.660209e+17      1.000000
25%      1292.50000  6.782670e+17    537.000000
50%      3229.00000  7.182469e+17   1239.000000
75%      9097.00000  7.986692e+17   3077.000000
max    153472.00000  8.924206e+17  76095.000000

In [18]:
tweet_data_df_clean.likes.sort_values()

453          0
439          0
437          0
434          0
431          0
         ...  
1055    113395
129     114540
517     118456
397     130823
1015    153472
Name: likes, Length: 2331, dtype: int64

In [19]:
tweet_data_df_clean.likes.value_counts().sort_values().tail(11)

110       2
3062      3
1362      3
2771      3
752       3
491       3
291       3
755       3
2033      3
3727      3
0       163
Name: likes, dtype: int64

In [20]:
tweet_data_df_clean.query("likes == 0")

timestamp  likes            tweet_id  retweets
31    Sat Jul 15 02:45:48 +0000 2017      0  886054160059072513        99
35    Thu Jul 13 01:35:06 +0000 2017      0  885311592912609280     16528
67    Mon Jun 26 00:13:58 +0000 2017      0  879130579576475649      6094
72    Sat Jun 24 00:09:53 +0000 2017      0  878404777348136964      1148
73    Fri Jun 23 18:17:33 +0000 2017      0  878316110768087041      5950
...                              ...    ...                 ...       ...
999   Sat Jun 25 01:52:36 +0000 2016      0  746521445350707200       966
1019  Fri Jun 17 16:01:16 +0000 2016      0  743835915802583040      2024
1218  Mon Mar 21 19:31:59 +0000 2016      0  711998809858043904       125
2234  Fri Nov 20 03:51:52 +0000 2015      0  667550904950915073        31
2235  Fri Nov 20 03:51:47 +0000 2015      0  667550882905632768        28

[163 rows x 4 columns]

In [21]:
tweet_data_df_clean.query("retweets == 1")

timestamp  likes            tweet_id  retweets
279  Sat Mar 04 17:56:49 +0000 2017    138  838085839343206401         1

In [22]:
tweet_data_df_clean.query("likes == 0 and retweets <= 20")

timestamp  likes            tweet_id  retweets
203  Tue Apr 11 18:15:55 +0000 2017      0  851861385021730816        19

In [23]:
tweet_data_df_clean.retweets.sort_values()

279         1
1271        2
262         2
328         2
109         3
        ...  
65      39888
397     42889
517     55468
1055    56601
1015    76095
Name: retweets, Length: 2331, dtype: int64

## image_predictions_clean

In [24]:
image_predictions_clean

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                      p1   p1_conf  p1_dog                  p2  \
0           1  Welsh_springer_spaniel  0.465074    True              collie   
1           1                 redbone  0.506826    True  miniature_pinscher   
2           1         German_shepherd  0.596461    True            malinois   
3           1     Rhodesian_ridgeback  0.408143    True             redbone   
4           1      miniature_pinscher  0.560311    True          Rottweiler   
...       ...                     ...       ...     ...                 ...   
2070        2                  basset  0.555712    True    English_springer   
2071        1             paper_towel  0.170278   False  Labrador_retriever   
2072        1               Chihuahua  0.716012    True            malamute   
2073        1               Chihuahua  0.323581    True            Pekinese   
2074        1                  orange  0.097049   False               bagel   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2071  0.168086    True                      spatula  0.040836   False  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  
2074  0.085851   False                       banana  0.076110   False  

[2075 rows x 12 columns]

In [25]:
image_predictions_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [26]:
image_predictions_clean.describe()

tweet_id      img_num      p1_conf       p2_conf       p3_conf
count  2.075000e+03  2075.000000  2075.000000  2.075000e+03  2.075000e+03
mean   7.384514e+17     1.203855     0.594548  1.345886e-01  6.032417e-02
std    6.785203e+16     0.561875     0.271174  1.006657e-01  5.090593e-02
min    6.660209e+17     1.000000     0.044333  1.011300e-08  1.740170e-10
25%    6.764835e+17     1.000000     0.364412  5.388625e-02  1.622240e-02
50%    7.119988e+17     1.000000     0.588230  1.181810e-01  4.944380e-02
75%    7.932034e+17     1.000000     0.843855  1.955655e-01  9.180755e-02
max    8.924206e+17     4.000000     1.000000  4.880140e-01  2.734190e-01

In [27]:
image_predictions_clean.p1.value_counts()

golden_retriever       150
Labrador_retriever     100
Pembroke                89
Chihuahua               83
pug                     57
                      ... 
orange                   1
black-footed_ferret      1
polecat                  1
boathouse                1
alp                      1
Name: p1, Length: 378, dtype: int64

In [28]:
image_predictions_clean.query("p1_dog == False and p2_dog == False and p3_dog == True")

tweet_id                                            jpg_url  \
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
87    667524857454854144    https://pbs.twimg.com/media/CUOGUfJW4AA_eni.jpg   
126   668256321989451776    https://pbs.twimg.com/media/CUYflCXWEAAzQVu.jpg   
131   668297328638447616    https://pbs.twimg.com/media/CUZE4IWW4AAZmDf.jpg   
158   668872652652679168    https://pbs.twimg.com/media/CUhQIAhXAAA2j7u.jpg   
...                  ...                                                ...   
1872  844979544864018432    https://pbs.twimg.com/media/C7n4aQ0VAAAohkL.jpg   
1899  851224888060895234    https://pbs.twimg.com/media/C9AohFoWsAUmxDs.jpg   
1942  861288531465048066  https://pbs.twimg.com/ext_tw_video_thumb/86128...   
1944  861769973181624320    https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg   
2026  882045870035918850    https://pbs.twimg.com/media/DD2oCl2WAAEI_4a.jpg   

      img_num             p1   p1_conf  p1_dog               p2   p2_conf  \
8           1  shopping_cart  0.962465   False  shopping_basket  0.014594   
87          1           hare  0.447893   False            dhole  0.092435   
126         1          canoe  0.407683   False           paddle  0.115550   
131         1   king_penguin  0.606747   False         ice_bear  0.264221   
158         1          teddy  0.413379   False           pillow  0.325623   
...       ...            ...       ...     ...              ...       ...   
1872        3    tennis_ball  0.999281   False           racket  0.000370   
1899        3     car_mirror  0.971512   False        seat_belt  0.007063   
1942        1        syringe  0.144712   False      oxygen_mask  0.106684   
1944        2  Arabian_camel  0.366248   False      house_finch  0.209852   
2026        1       web_site  0.949591   False            dhole  0.017326   

      p2_dog                        p3   p3_conf  p3_dog  
8      False          golden_retriever  0.007959    True  
87     False  Chesapeake_Bay_retriever  0.088122    True  
126    False                  Pembroke  0.094429    True  
131    False                Eskimo_dog  0.032784    True  
158    False       miniature_schnauzer  0.035537    True  
...      ...                       ...       ...     ...  
1872   False         Shetland_sheepdog  0.000132    True  
1899   False           standard_poodle  0.005683    True  
1942   False      Bouvier_des_Flandres  0.082610    True  
1944   False            cocker_spaniel  0.046403    True  
2026   False          golden_retriever  0.006941    True  

[64 rows x 12 columns]

In [29]:
image_predictions_clean.query("p1_dog == False and p2_dog == False and p3_dog == False")

tweet_id                                            jpg_url  \
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
...                  ...                                                ...   
2021  880935762899988482    https://pbs.twimg.com/media/DDm2Z5aXUAEDS2u.jpg   
2022  881268444196462592    https://pbs.twimg.com/media/DDrk-f9WAAI-WQv.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2074  892420643555336193    https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num                p1   p1_conf  p1_dog               p2   p2_conf  \
6           1        box_turtle  0.933012   False       mud_turtle  0.045885   
17          1               hen  0.965932   False             cock  0.033919   
18          1  desktop_computer  0.086502   False             desk  0.085547   
21          1  three-toed_sloth  0.914671   False            otter  0.015250   
25          1        guinea_pig  0.996496   False            skunk  0.002402   
...       ...               ...       ...     ...              ...       ...   
2021        1       street_sign  0.251801   False         umbrella  0.115123   
2022        1            tusker  0.473303   False  Indian_elephant  0.245646   
2046        1       convertible  0.738995   False       sports_car  0.139952   
2052        1         limousine  0.130432   False        tow_truck  0.029175   
2074        1            orange  0.097049   False            bagel  0.085851   

      p2_dog              p3   p3_conf  p3_dog  
6      False        terrapin  0.017885   False  
17     False       partridge  0.000052   False  
18     False        bookcase  0.079480   False  
21     False  great_grey_owl  0.013207   False  
25     False         hamster  0.000461   False  
...      ...             ...       ...     ...  
2021   False   traffic_light  0.069534   False  
2022   False            ibex  0.055661   False  
2046   False       car_wheel  0.044173   False  
2052   False   shopping_cart  0.026321   False  
2074   False          banana  0.076110   False  

[324 rows x 12 columns]

In [30]:
image_predictions_clean.query("p1_dog == False and p2_dog == True and p3_dog == True")

tweet_id                                            jpg_url  \
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
43    666776908487630848    https://pbs.twimg.com/media/CUDeDoWUYAAD-EM.jpg   
95    667550904950915073    https://pbs.twimg.com/media/CUOb_gUUkAACXdS.jpg   
103   667806454573760512    https://pbs.twimg.com/media/CUSGbXeVAAAgztZ.jpg   
108   667878741721415682    https://pbs.twimg.com/media/CUTILFiWcAE8Rle.jpg   
...                  ...                                                ...   
1858  841833993020538882  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1896  850145622816686080    https://pbs.twimg.com/media/C8xS655XkAAv9vo.jpg   
1904  852189679701164033    https://pbs.twimg.com/media/C9OV99SXsAEmj1U.jpg   
1984  872122724285648897    https://pbs.twimg.com/media/DBpm-5UXcAUeCru.jpg   
2013  879376492567855104    https://pbs.twimg.com/media/DDQsQGFV0AAw6u9.jpg   

      img_num           p1   p1_conf  p1_dog                    p2   p2_conf  \
22          1           ox  0.416669   False          Newfoundland  0.278407   
43          1    seat_belt  0.375057   False    miniature_pinscher  0.167175   
95          1     web_site  0.999335   False                vizsla  0.000081   
103         1      toyshop  0.253089   False             Chihuahua  0.187155   
108         1    seat_belt  0.200373   False    miniature_pinscher  0.106003   
...       ...          ...       ...     ...                   ...       ...   
1858        1     ice_bear  0.336200   False               Samoyed  0.201358   
1896        2  tennis_ball  0.714798   False                kelpie  0.105390   
1904        1       barrow  0.423150   False  Bernese_mountain_dog  0.415374   
1984        1   basketball  0.808396   False                   pug  0.066736   
2013        1     tricycle  0.663601   False    Labrador_retriever  0.033496   

      p2_dog                 p3   p3_conf  p3_dog  
22      True        groenendael  0.102643    True  
43      True          Chihuahua  0.086951    True  
95      True             collie  0.000069    True  
103     True  Brabancon_griffon  0.112799    True  
108     True         schipperke  0.104733    True  
...      ...                ...       ...     ...  
1858    True         Eskimo_dog  0.186789    True  
1896    True           malinois  0.058553    True  
1904    True        EntleBucher  0.067345    True  
1984    True          dalmatian  0.054570    True  
2013    True           Pembroke  0.018827    True  

[101 rows x 12 columns]

In [31]:
image_predictions_clean.query("p1_dog == True")

tweet_id                                          jpg_url  \
0     666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
...                  ...                                              ...   
2068  890971913173991426  https://pbs.twimg.com/media/DF1eOmZXUAALUcq.jpg   
2069  891087950875897856  https://pbs.twimg.com/media/DF3HwyEWsAABqE6.jpg   
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   

      img_num                        p1   p1_conf  p1_dog                  p2  \
0           1    Welsh_springer_spaniel  0.465074    True              collie   
1           1                   redbone  0.506826    True  miniature_pinscher   
2           1           German_shepherd  0.596461    True            malinois   
3           1       Rhodesian_ridgeback  0.408143    True             redbone   
4           1        miniature_pinscher  0.560311    True          Rottweiler   
...       ...                       ...       ...     ...                 ...   
2068        1               Appenzeller  0.341703    True       Border_collie   
2069        1  Chesapeake_Bay_retriever  0.425595    True       Irish_terrier   
2070        2                    basset  0.555712    True    English_springer   
2072        1                 Chihuahua  0.716012    True            malamute   
2073        1                 Chihuahua  0.323581    True            Pekinese   

       p2_conf  p2_dog                           p3   p3_conf  p3_dog  
0     0.156665    True            Shetland_sheepdog  0.061428    True  
1     0.074192    True          Rhodesian_ridgeback  0.072010    True  
2     0.138584    True                   bloodhound  0.116197    True  
3     0.360687    True           miniature_pinscher  0.222752    True  
4     0.243682    True                     Doberman  0.154629    True  
...        ...     ...                          ...       ...     ...  
2068  0.199287    True                    ice_lolly  0.193548   False  
2069  0.116317    True              Indian_elephant  0.076902   False  
2070  0.225770    True  German_short-haired_pointer  0.175219    True  
2072  0.078253    True                       kelpie  0.031379    True  
2073  0.090647    True                     papillon  0.068957    True  

[1532 rows x 12 columns]

## twitter_archive_clean

In [90]:
twitter_archive_clean.sample(3)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
435   820446719150292993                    NaN                  NaN   
2006  672267570918129665                    NaN                  NaN   
928   754747087846248448                    NaN                  NaN   

                      timestamp  \
435   2017-01-15 01:45:15 +0000   
2006  2015-12-03 04:14:13 +0000   
928   2016-07-17 18:38:22 +0000   

                                                                                  source  \
435   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
2006  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
928   <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                         text  \
435   RT @dog_rates: This is Peaches. She's the ultimate selfie sidekick. Super sneaky tongue slip appreciated. 13/10 https://t.co/pbKOesr8Tg   
2006                                                     When you accidentally open up the front facing camera. 10/10 https://t.co/jDXxZARQIZ   
928                This is Keith. He's pursuing a more 2D lifestyle. Idiosyncratic af. 12/10 follow your dreams Keith https://t.co/G9ufksBMlU   

      retweeted_status_id  retweeted_status_user_id  \
435          8.001414e+17              4.196984e+09   
2006                  NaN                       NaN   
928                   NaN                       NaN   

     retweeted_status_timestamp  \
435   2016-11-20 00:59:15 +0000   
2006                        NaN   
928                         NaN   

                                                                                                                                                                                                                                                                                                                                                                                        expanded_urls  \
435   https://twitter.com/dog_rates/status/800141422401830912/photo/1,https://twitter.com/dog_rates/status/800141422401830912/photo/1,https://twitter.com/dog_rates/status/800141422401830912/photo/1,https://twitter.com/dog_rates/status/800141422401830912/photo/1,https://twitter.com/dog_rates/status/800141422401830912/photo/1,https://twitter.com/dog_rates/status/800141422401830912/photo/1   
2006                                                                                                                                                                                                                                                                                                                                  https://twitter.com/dog_rates/status/672267570918129665/photo/1   
928                                                                                                                                                                                                                                                                                                                                   https://twitter.com/dog_rates/status/754747087846248448/photo/1   

      rating_numerator  rating_denominator     name doggo floofer pupper puppo  
435                 13                  10  Peaches  None    None   None  None  
2006                10                  10     None  None    None   None  None  
928                 12                  10    Keith  None    None   None  None

In [33]:
twitter_archive_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [34]:
twitter_archive_clean.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [69]:
twitter_archive_clean.duplicated().value_counts()

False    2356
dtype: int64

In [70]:
2356 - 2297

59

In [95]:
twitter_archive_clean.query("rating_denominator == 0")

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
313  835246439529840640           8.352460e+17           26259576.0   

                     timestamp  \
313  2017-02-24 21:54:03 +0000   

                                                                                 source  \
313  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                   text  \
313  @jonnysun @Lin_Manuel ok jomny I know you're excited but 960/00 isn't a valid rating, 13/10 is tho   

     retweeted_status_id  retweeted_status_user_id retweeted_status_timestamp  \
313                  NaN                       NaN                        NaN   

    expanded_urls  rating_numerator  rating_denominator  name doggo floofer  \
313           NaN               960                   0  None  None    None   

    pupper puppo  
313   None  None